File: microfreak/collections.py
"""PresetCollection: a named device arrangement — an ordered slot->PresetRef
map over the library's content-addressed blobs, plus identity and provenance.

This module is pure of device I/O: it defines the collection value types, their
JSON codec, and the pure apply/switch PLAN (plan_apply). It imports model and
errors ONLY (never device or library), so planning is importable everywhere and
stays free of any wire/byte dependency. Executing a plan is
MicroFreak.apply_collection (device.py); building/storing collections is the
Library (library.py).

On-disk format — one file per collection under the library:

    <root>/collections/<id>.json
      {"schema": 1, "id": str, "name": str, "created_at": ISO8601,
       "provenance": {"kind": str, "source": str},
       "slots": {"<slot>": {"sha256": str, "name": str, "meta_hex": str}}}

`slots` is keyed by the DECIMAL slot number as a string (mirrors the backup
`presets` map). Canonical iteration is ascending by int(key); JSON key order is
irrelevant. Referenced blobs live in the SHARED blobs/ dir; a collection stores
no blob bytes of its own.
"""
from __future__ import annotations

import enum
import time
import uuid
from dataclasses import dataclass
from typing import Dict, Optional, Tuple

from .errors import LibraryCorruptError
from .model import DeviceSnapshot, PresetRef, SlotRecord

_COLLECTION_SCHEMA = 1
_ZERO_META_HEX = "00" * 9        # a valid, writable all-zero meta (Ambient Peaks)


# --------------------------------------------------------------- provenance

class ProvenanceKind(enum.Enum):
    DEVICE_SNAPSHOT = "device_snapshot"
    IMPORTED_BANK = "imported_bank"
    MANUAL = "manual"

    @classmethod
    def from_str(cls, s: str) -> "ProvenanceKind":
        try:
            return cls(s)
        except ValueError:
            return cls.MANUAL


@dataclass(frozen=True)
class Provenance:
    kind: ProvenanceKind
    source: str = ""            # filename / snapshot taken_at / ""


@dataclass(frozen=True)
class PresetCollection:
    id: str                          # uuid4 hex, minted at creation
    name: str
    created_at: str                  # ISO 8601, local
    provenance: Provenance
    slots: Dict[int, PresetRef]      # slot -> ref; iterate sorted(slots)

    @classmethod
    def new(cls, name: str, provenance: Provenance,
            slots: Dict[int, PresetRef]) -> "PresetCollection":
        return cls(id=uuid.uuid4().hex, name=name,
                   created_at=time.strftime("%Y-%m-%dT%H:%M:%S"),
                   provenance=provenance, slots=dict(slots))

    def covered_slots(self) -> Tuple[int, ...]:
        return tuple(sorted(self.slots))


# ---------------------------------------------------------------- bank input

@dataclass(frozen=True)
class BankItem:
    """Core input value type for import-.mfprojz-as-Collection. The caller
    (orchestrator/app) runs the verified tools/mbp_import.py parser and adapts
    each MbpPreset to a BankItem:
        BankItem(slot=p.slot, name=p.name, meta=p.meta, blob=p.blob)
    keeping the core stdlib-clean and free of a tools/ dependency."""
    slot: Optional[int]     # 0-based MF slot from the filename, or None
    name: str
    meta: bytes             # 9 bytes, or b"" for an empty/Init-only slot
    blob: Optional[bytes]   # 4672 bytes, or None for an empty slot


# ------------------------------------------------------------------- codec

def _ref_to_json(ref: PresetRef) -> dict:
    return {"sha256": ref.sha256, "name": ref.name, "meta_hex": ref.meta_hex}


def _ref_from_json(d: dict) -> PresetRef:
    return PresetRef(sha256=d["sha256"], name=d["name"], meta_hex=d["meta_hex"])


def collection_to_json(coll: PresetCollection) -> dict:
    return {
        "schema": _COLLECTION_SCHEMA,
        "id": coll.id,
        "name": coll.name,
        "created_at": coll.created_at,
        "provenance": {"kind": coll.provenance.kind.value,
                       "source": coll.provenance.source},
        "slots": {str(slot): _ref_to_json(coll.slots[slot])
                  for slot in sorted(coll.slots)},
    }


def collection_from_json(d: dict, *, path: str = "<collection>") -> PresetCollection:
    if d.get("schema") != _COLLECTION_SCHEMA:
        raise LibraryCorruptError(path, f"unsupported schema: {d.get('schema')!r}")
    try:
        prov_d = d.get("provenance") or {}
        provenance = Provenance(
            kind=ProvenanceKind.from_str(prov_d.get("kind", "manual")),
            source=prov_d.get("source", ""))
        slots: Dict[int, PresetRef] = {
            int(k): _ref_from_json(v) for k, v in (d.get("slots") or {}).items()}
        return PresetCollection(id=d["id"], name=d["name"],
                                created_at=d.get("created_at", ""),
                                provenance=provenance, slots=slots)
    except (KeyError, TypeError, ValueError) as e:
        raise LibraryCorruptError(path, f"bad collection: {e}") from e


# ---------------------------------------------------------- apply / switch plan

class PlanAction(enum.Enum):
    WRITE = "write"            # content or name differs -> full verified write
    SKIP_UNCHANGED = "skip"    # device already matches (sha AND name equal)
    CLEAR = "clear"            # not in collection; overwrite with clear_with


@dataclass(frozen=True)
class ApplyOptions:
    unlisted: str = "leave"          # "leave" | "clear" — slots absent from the collection
    clear_with: Optional[PresetRef] = None   # REQUIRED when unlisted == "clear"
    seconds_per_write: float = 1.0   # verified-write estimate (~0.5 s write + ~0.4 s verify)


@dataclass(frozen=True)
class SlotPlan:
    slot: int
    action: PlanAction
    incoming: Optional[PresetRef]    # ref to write (WRITE/CLEAR); None for SKIP
    victim: Optional[SlotRecord]     # the device record being replaced (for UI framing)


@dataclass(frozen=True)
class ApplyPlan:
    slots: Tuple[SlotPlan, ...]      # one per device slot, ascending
    write_count: int
    clear_count: int
    skip_count: int
    total_slots: int                 # == len(slots) (the device's slot count)
    estimated_seconds: float         # round((write+clear) * seconds_per_write, 1)

    def changes(self) -> Tuple[SlotPlan, ...]:
        return tuple(p for p in self.slots
                     if p.action in (PlanAction.WRITE, PlanAction.CLEAR))


def plan_apply(collection: PresetCollection, snapshot: DeviceSnapshot, *,
               options: ApplyOptions = ApplyOptions()) -> ApplyPlan:
    """Pure. Requires a FULL hashed snapshot: every slot 0..N-1 present and
    snapshot.has_hashes, else ValueError. If options.unlisted == 'clear',
    options.clear_with is required, else ValueError. A collection slot beyond
    the snapshot's range makes the plan undecidable -> ValueError.

    Per device record (ascending), ref = collection.slots.get(slot):
      ref present:
        record.sha256 == ref.sha256 and record.name == ref.name -> SKIP_UNCHANGED
        else                                                      -> WRITE
      ref absent:
        unlisted == 'clear':
          record already equals clear_with (sha AND name) -> SKIP_UNCHANGED
          else                                            -> CLEAR
        unlisted == 'leave':                              -> SKIP_UNCHANGED
    """
    records = sorted(snapshot.records, key=lambda r: r.slot)
    total = len(records)
    slot_set = {r.slot for r in records}
    if slot_set != set(range(total)):
        raise ValueError(
            "plan_apply requires a FULL snapshot: every slot 0..N-1 present")
    if not snapshot.has_hashes:
        raise ValueError("plan_apply requires a snapshot with blob hashes")
    if options.unlisted not in ("leave", "clear"):
        raise ValueError(f"unknown unlisted policy: {options.unlisted!r}")
    if options.unlisted == "clear" and options.clear_with is None:
        raise ValueError("unlisted == 'clear' requires options.clear_with")
    if collection.slots and max(collection.slots) >= total:
        raise ValueError(
            f"collection references slot {max(collection.slots)} beyond the "
            f"snapshot's {total} slots")

    plans = []
    write_count = clear_count = skip_count = 0
    for r in records:
        ref = collection.slots.get(r.slot)
        if ref is not None:
            if r.sha256 == ref.sha256 and r.name == ref.name:
                plans.append(SlotPlan(r.slot, PlanAction.SKIP_UNCHANGED, None, None))
                skip_count += 1
            else:
                plans.append(SlotPlan(r.slot, PlanAction.WRITE, ref, r))
                write_count += 1
        elif options.unlisted == "clear":
            cw = options.clear_with
            if r.sha256 == cw.sha256 and r.name == cw.name:
                plans.append(SlotPlan(r.slot, PlanAction.SKIP_UNCHANGED, None, None))
                skip_count += 1
            else:
                plans.append(SlotPlan(r.slot, PlanAction.CLEAR, cw, r))
                clear_count += 1
        else:   # leave
            plans.append(SlotPlan(r.slot, PlanAction.SKIP_UNCHANGED, None, None))
            skip_count += 1

    estimated = round((write_count + clear_count) * options.seconds_per_write, 1)
    return ApplyPlan(slots=tuple(plans), write_count=write_count,
                     clear_count=clear_count, skip_count=skip_count,
                     total_slots=total, estimated_seconds=estimated)

File: ios/FreakCore/Sources/FreakCore/Collections.swift
// Collections.swift — PresetCollection: a named device arrangement, an
// ordered slot->PresetRef map over the library's content-addressed blobs,
// plus identity and provenance. Port of microfreak/collections.py.
//
// This file is pure of device I/O: it defines the collection value types,
// their JSON codec, and the pure apply/switch PLAN (planApply). Building and
// storing collections is the Library actor (Library.swift); executing a plan
// is MicroFreakDevice.applyCollection (Device.swift).
//
// On-disk format — one file per collection under the library:
//
//     <root>/collections/<id>.json
//       {"schema": 1, "id": str, "name": str, "created_at": ISO8601,
//        "provenance": {"kind": str, "source": str},
//        "slots": {"<slot>": {"sha256": str, "name": str, "meta_hex": str}}}
//
// `slots` is keyed by the DECIMAL slot number as a string (mirrors the backup
// `presets` map). Canonical iteration is ascending by int(key); JSON key order
// is irrelevant. Referenced blobs live in the SHARED blobs/ dir; a collection
// stores no blob bytes of its own.

import Foundation

let collectionSchema = 1
private let zeroMetaHex = String(repeating: "00", count: 9)  // valid all-zero meta

// --------------------------------------------------------------- provenance

public enum ProvenanceKind: String, Sendable, Codable, Equatable {
    case deviceSnapshot = "device_snapshot"
    case importedBank   = "imported_bank"
    case manual

    /// Parse a file value. Unknown -> .manual (forward compatibility).
    public static func fromString(_ s: String) -> ProvenanceKind {
        ProvenanceKind(rawValue: s) ?? .manual
    }
}

public struct Provenance: Sendable, Equatable {
    public let kind: ProvenanceKind
    public let source: String            // filename / snapshot takenAt / ""

    public init(kind: ProvenanceKind, source: String = "") {
        self.kind = kind
        self.source = source
    }
}

public struct PresetCollection: Sendable, Equatable, Identifiable {
    public let id: String                // uuid4 hex, minted at creation
    public let name: String
    public let createdAt: String         // ISO 8601, local
    public let provenance: Provenance
    public let slots: [Int: PresetRef]   // slot -> ref; iterate slots.keys.sorted()

    public init(id: String, name: String, createdAt: String,
                provenance: Provenance, slots: [Int: PresetRef]) {
        self.id = id
        self.name = name
        self.createdAt = createdAt
        self.provenance = provenance
        self.slots = slots
    }

    /// Mint a fresh collection (uuid4 hex id, local ISO createdAt).
    public static func new(name: String, provenance: Provenance,
                           slots: [Int: PresetRef]) -> PresetCollection {
        PresetCollection(
            id: UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased(),
            name: name, createdAt: isoNow(), provenance: provenance, slots: slots)
    }

    public func coveredSlots() -> [Int] { slots.keys.sorted() }

    /// Copy with a new name (rename keeps id, createdAt, provenance, slots).
    func renamed(_ name: String) -> PresetCollection {
        PresetCollection(id: id, name: name, createdAt: createdAt,
                         provenance: provenance, slots: slots)
    }
}

// ---------------------------------------------------------------- bank input

/// Core input value type for import-.mfprojz-as-Collection. The caller
/// (orchestrator/app) runs the verified tools/mbp_import.py parser (or a Swift
/// .mfprojz reader) and adapts each parsed preset to a BankItem, keeping the
/// core free of a Boost-archive dependency.
public struct BankItem: Sendable, Equatable {
    public let slot: Int?        // 0-based MF slot from the filename, or nil
    public let name: String
    public let meta: Data        // 9 bytes, or empty for an empty/Init-only slot
    public let blob: Data?       // 4672 bytes, or nil for an empty slot

    public init(slot: Int?, name: String, meta: Data, blob: Data?) {
        self.slot = slot
        self.name = name
        self.meta = meta
        self.blob = blob
    }
}

// ------------------------------------------------------------------- codec

enum CollectionCodec {
    static func toJSON(_ coll: PresetCollection) -> [String: Any] {
        var slotsOut: [String: Any] = [:]
        for slot in coll.slots.keys.sorted() {
            let ref = coll.slots[slot]!
            slotsOut[String(slot)] = ["sha256": ref.sha256, "name": ref.name,
                                      "meta_hex": ref.metaHex]
        }
        return [
            "schema": collectionSchema,
            "id": coll.id,
            "name": coll.name,
            "created_at": coll.createdAt,
            "provenance": ["kind": coll.provenance.kind.rawValue,
                           "source": coll.provenance.source],
            "slots": slotsOut,
        ]
    }

    /// Parse a collection dictionary. An unparseable or schema != 1 document
    /// throws .libraryCorrupt(path:, detail:).
    static func fromJSON(_ d: [String: Any], path: String) throws -> PresetCollection {
        guard (d["schema"] as? NSNumber)?.intValue == collectionSchema else {
            throw FreakError.libraryCorrupt(
                path: path, detail: "unsupported schema: \(d["schema"] ?? "nil")")
        }
        guard let id = d["id"] as? String, let name = d["name"] as? String else {
            throw FreakError.libraryCorrupt(path: path, detail: "bad collection: missing id/name")
        }
        let provD = d["provenance"] as? [String: Any] ?? [:]
        let provenance = Provenance(
            kind: ProvenanceKind.fromString(provD["kind"] as? String ?? "manual"),
            source: provD["source"] as? String ?? "")
        var slots: [Int: PresetRef] = [:]
        for (k, v) in (d["slots"] as? [String: Any] ?? [:]) {
            guard let slot = Int(k), let rd = v as? [String: Any],
                  let sha = rd["sha256"] as? String,
                  let rname = rd["name"] as? String,
                  let metaHex = rd["meta_hex"] as? String else {
                throw FreakError.libraryCorrupt(path: path, detail: "bad slot ref: \(k)")
            }
            slots[slot] = PresetRef(sha256: sha, name: rname, metaHex: metaHex)
        }
        return PresetCollection(id: id, name: name,
                                createdAt: d["created_at"] as? String ?? "",
                                provenance: provenance, slots: slots)
    }
}

// ---------------------------------------------------------- apply / switch plan

public enum PlanAction: String, Sendable, Equatable {
    case write                        // content OR name differs -> full verified write
    case skip = "skip"                // device already matches (sha AND name equal)
    case clear                        // not in collection; overwrite with clearWith
}

public struct ApplyOptions: Sendable {
    public enum Unlisted: Sendable, Equatable { case leave, clear }
    public var unlisted: Unlisted = .leave
    public var clearWith: PresetRef? = nil      // REQUIRED when unlisted == .clear
    public var secondsPerWrite: Double = 1.0    // ~0.5 s write + ~0.4 s verify

    public init() {}
}

public struct SlotPlan: Sendable, Equatable {
    public let slot: Int
    public let action: PlanAction
    public let incoming: PresetRef?    // ref to write (WRITE/CLEAR); nil for SKIP
    public let victim: SlotRecord?     // the device record being replaced (UI framing)

    public init(slot: Int, action: PlanAction, incoming: PresetRef?, victim: SlotRecord?) {
        self.slot = slot
        self.action = action
        self.incoming = incoming
        self.victim = victim
    }
}

public struct ApplyPlan: Sendable, Equatable {
    public let slots: [SlotPlan]       // one per device slot, ascending
    public let writeCount: Int
    public let clearCount: Int
    public let skipCount: Int
    public let totalSlots: Int         // == slots.count (the device's slot count)
    public let estimatedSeconds: Double

    public init(slots: [SlotPlan], writeCount: Int, clearCount: Int,
                skipCount: Int, totalSlots: Int, estimatedSeconds: Double) {
        self.slots = slots
        self.writeCount = writeCount
        self.clearCount = clearCount
        self.skipCount = skipCount
        self.totalSlots = totalSlots
        self.estimatedSeconds = estimatedSeconds
    }

    /// The changed slots only (WRITE + CLEAR) — what applyCollection writes.
    public func changes() -> [SlotPlan] {
        slots.filter { $0.action == .write || $0.action == .clear }
    }
}

/// Pure. Requires a FULL hashed snapshot: every slot 0..N-1 present and
/// snapshot.hasHashes, else it throws. If options.unlisted == .clear,
/// options.clearWith is required, else it throws. A collection slot beyond the
/// snapshot's range makes the plan undecidable and throws.
///
/// Per device record (ascending), ref = collection.slots[slot]:
///   ref present:
///     record.sha256 == ref.sha256 && record.name == ref.name -> SKIP_UNCHANGED
///     else                                                     -> WRITE (incoming=ref)
///   ref absent:
///     unlisted == .clear:
///       record already equals clearWith (sha AND name) -> SKIP_UNCHANGED
///       else                                           -> CLEAR (incoming=clearWith)
///     unlisted == .leave:                              -> SKIP_UNCHANGED
///
/// A name-only difference (sha equal, name differing) is a WRITE, folded into
/// the full verified write for the three-action model.
public func planApply(collection: PresetCollection, snapshot: DeviceSnapshot,
                      options: ApplyOptions = ApplyOptions()) throws -> ApplyPlan {
    let records = snapshot.records.sorted { $0.slot < $1.slot }
    let total = records.count
    let slotSet = Set(records.map(\.slot))
    guard slotSet == Set(0..<total) else {
        throw FreakError.protocolViolation(
            detail: "planApply requires a FULL snapshot: every slot 0..N-1 present")
    }
    guard snapshot.hasHashes else {
        throw FreakError.snapshotMissingHashes
    }
    if options.unlisted == .clear, options.clearWith == nil {
        throw FreakError.protocolViolation(
            detail: "unlisted == .clear requires options.clearWith")
    }
    if let maxSlot = collection.slots.keys.max(), maxSlot >= total {
        throw FreakError.protocolViolation(
            detail: "collection references slot \(maxSlot) beyond the "
                + "snapshot's \(total) slots")
    }

    var plans: [SlotPlan] = []
    var writeCount = 0, clearCount = 0, skipCount = 0
    for r in records {
        if let ref = collection.slots[r.slot] {
            if r.sha256 == ref.sha256 && r.name == ref.name {
                plans.append(SlotPlan(slot: r.slot, action: .skip, incoming: nil, victim: nil))
                skipCount += 1
            } else {
                plans.append(SlotPlan(slot: r.slot, action: .write, incoming: ref, victim: r))
                writeCount += 1
            }
        } else if options.unlisted == .clear {
            let cw = options.clearWith!
            if r.sha256 == cw.sha256 && r.name == cw.name {
                plans.append(SlotPlan(slot: r.slot, action: .skip, incoming: nil, victim: nil))
                skipCount += 1
            } else {
                plans.append(SlotPlan(slot: r.slot, action: .clear, incoming: cw, victim: r))
                clearCount += 1
            }
        } else {   // leave
            plans.append(SlotPlan(slot: r.slot, action: .skip, incoming: nil, victim: nil))
            skipCount += 1
        }
    }

    let estimated = roundTo(Double(writeCount + clearCount) * options.secondsPerWrite,
                            places: 1)
    return ApplyPlan(slots: plans, writeCount: writeCount, clearCount: clearCount,
                     skipCount: skipCount, totalSlots: total, estimatedSeconds: estimated)
}
